<a href="https://colab.research.google.com/github/sgathai/dsc-data-serialization-lab/blob/master/Copy_of_Phase_3_Code_Challenge_Notebook_ipynb_Sgathai.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Phase 3 Code Challenge

This assessment is designed to test your understanding of Module 3 material. It covers:

* Gradient Descent
* Logistic Regression
* Classification Metrics
* Decision Trees

_Read the instructions carefully_. You will be asked both to write code and to answer short answer questions.

## Code Tests

We have provided some code tests for you to run to check that your work meets the item specifications. Passing these tests does not necessarily mean that you have gotten the item correct - there are additional hidden tests. However, if any of the tests do not pass, this tells you that your code is incorrect and needs changes to meet the specification. To determine what the issue is, read the comments in the code test cells, the error message you receive, and the item instructions.

## Short Answer Questions

For the short answer questions...

* _Use your own words_. It is OK to refer to outside resources when crafting your response, but _do not copy text from another source_.

* _Communicate clearly_. We are not grading your writing skills, but you can only receive full credit if your teacher is able to fully understand your response.

* _Be concise_. You should be able to answer most short answer questions in a sentence or two. Writing unnecessarily long answers increases the risk of you being unclear or saying something incorrect.

In [ ]:
# Run this cell without changes to import the necessary libraries

from numbers import Number

%matplotlib inline

---
## Part 1: Gradient Descent [Suggested Time: 20 min]
---
In this part, you will describe how gradient descent works to calculate a parameter estimate. Below is an image of a best fit line from a linear regression model using TV advertising spending to predict product sales.

![best fit line](https://raw.githubusercontent.com/learn-co-curriculum/dsc-cc-images/main/phase_3/best_fit_line.png)

This best fit line can be described by the equation $y = mx + b$. Below is the RSS cost curve associated with the slope parameter $m$:

![cost curve](https://raw.githubusercontent.com/learn-co-curriculum/dsc-cc-images/main/phase_3/cost_curve.png)

where RSS is the residual sum of squares: $RSS = \sum_{i=1}^n(y_i - (mx_i + b))^2$

### 1.1) Short Answer: Explain how the RSS curve above could be used to find an optimal value for the slope parameter $m$.

Your answer should provide a one sentence summary, not every step of the process.

***By finding in the rss axises the point where  slope has the least value which is the minimal error in line with lowest value in Slope (m-values ).***

Below is a visualization showing the iterations of a gradient descent algorithm applied the RSS curve. Each yellow marker represents an estimate, and the lines between markers represent the steps taken between estimates in each iteration. Numeric labels identify the iteration numbers.

![gradient descent](https://raw.githubusercontent.com/learn-co-curriculum/dsc-cc-images/main/phase_3/gd.png)

### 1.2) Short Answer: Explain why the distances between markers get smaller over successive iterations.

**The steps gets smaller so that we  minimize chances of missing the optimal value of parameter (m)**

### 1.3) Short Answer: What would be the effect of decreasing the learning rate for this application of gradient descent?

**This will increase the steps done but will help achieve the optimal value of parameter  (m)**

---
## Part 2: Logistic Regression [Suggested Time: 15 min]
---
In this part, you will answer general questions about logistic regression.

### 2.1) Short Answer: Provide one reason why logistic regression is better than linear regression for modeling a binary target/outcome.

**Logistics regression unlike linear regression takes into account the probability of a binary/categorical variable outcome or target  .**

### 2.2) Short Answer: Compare logistic regression to another classification model of your choice (e.g. KNN, Decision Tree, etc.). What is one advantage and one disadvantage logistic regression has when compared with the other model?

The advantage of logistics regression is that it  handles multicollineariality ,can handle both continous and categorical predictors thus enhancing  clearerer  understanding of the likelihood of outcomes unlike linear regression and KNN.

The disadvantage of logistics regression is that it assumes linearity which at times may not hold

---
## Part 3: Classification Metrics [Suggested Time: 20 min]
---
In this part, you will make sense of classification metrics produced by various classifiers.

The confusion matrix below represents the predictions generated by a classisification model on a small testing dataset.

![cnf matrix](https://raw.githubusercontent.com/learn-co-curriculum/dsc-cc-images/main/phase_3/cnf_matrix.png)

### 3.1) Create a numeric variable `precision` containing the precision of the classifier.

**Starter Code**

    precision =true positive/(true positive+true negative)

In [ ]:
# YOUR CODE HERE
#raise NotImplementedError()
#precision =true positive/(true positive+true negative)
true_positive=30
true_negative=4

precision=(30/(30+4))
precision


0.8823529411764706

In [ ]:
# This test confirms that you have created a numeric variable named precision

assert isinstance(precision, Number)


### 3.2) Create a numeric variable `f1score` containing the F-1 score of the classifier.

**Starter Code**

    f1score =2TP/(2TP+FP+FN)

In [ ]:
# YOUR CODE HERE
#raise NotImplementedError()
#recall=true positive/(true positive+false negative)
recall=30/(30+12)
f1score=(2*recall*precision)/(recall+precision)
f1score



0.7894736842105262

In [ ]:
# This test confirms that you have created a numeric variable named f1score

assert isinstance(f1score, Number)


The ROC curves below were calculated for three different models applied to one dataset.

1. Only Age was used as a feature in the model
2. Only Estimated Salary was used as a feature in the model
3. All features were used in the model

![roc](https://raw.githubusercontent.com/learn-co-curriculum/dsc-cc-images/main/phase_3/many_roc.png)

### 3.3) Short Answer: Identify the best ROC curve in the above graph and explain why it is the best.

All features ROC curve has a higher area under the curve as compared to the  others (age and estimated salary).Also using various features allows  the model to have a combined effect of multiple predictors .

Run the following cells to load a sample dataset, run a classification model on it, and perform some EDA.

In [ ]:
# Run this cell without changes

# Include relevant imports
import pickle, sklearn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_curve, roc_auc_score

network_df = pickle.load(open('sample_network_data.pkl', 'rb'))

# partion features and target
X = network_df.drop('Purchased', axis=1)
y = network_df['Purchased']

# train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2019)

# scale features
scale = StandardScaler()
scale.fit(X_train)
X_train = scale.transform(X_train)
X_test = scale.transform(X_test)

# build classifier
model = LogisticRegression(C=1e5, solver='lbfgs')
model.fit(X_train, y_train)
y_test_pred = model.predict(X_test)

# get the accuracy score
print(f'The classifier has an accuracy score of {round(accuracy_score(y_test, y_test_pred), 3)}.')

The classifier has an accuracy score of 0.956.


In [ ]:
# Run this cell without changes

y.value_counts()

0    257
1     13
Name: Purchased, dtype: int64

### 3.4) Short Answer: Explain how the distribution of `y` shown above could explain the high accuracy score of the classification model.

The distribution has a class imbalance where most of the instances the model would predict class 0 .This would be okay resulting to a higher score.

### 3.5) Short Answer: What is one method you could use to improve your model to address the issue discovered in Question 3.4?

undersampling

---
## Part 4: Decision Trees [Suggested Time: 20 min]
---
In this part, you will use decision trees to fit a classification model to a wine dataset. The data contain the results of a chemical analysis of wines grown in one region in Italy using three different cultivars (grape types). There are thirteen features from the measurements taken, and the wines are classified by cultivar in the `target` variable.

In [ ]:
# Run this cell without changes

# Relevant imports
import pandas as pd
import numpy as np
from sklearn.datasets import load_wine
from sklearn.tree import DecisionTreeClassifier

# Load the data
wine = load_wine()
X, y = load_wine(return_X_y=True)
X = pd.DataFrame(X, columns=wine.feature_names)
y = pd.Series(y)
y.name = 'target'

### 4.1) Use `train_test_split()` to evenly split `X` and `y` data between training sets (`X_train` and `y_train`) and test sets (`X_test` and `y_test`), with `random_state=1`.

Do not alter `X` or `y` before performing the split.

**Starter Code**

    X_train, X_test, y_train, y_test =

In [ ]:
# YOUR CODE HERE
#raise NotImplementedError()
X_train, X_test, y_train, y_test =train_test_split(X, y, test_size=0.5, random_state=1)




In [ ]:
# These tests confirm that you have created DataFrames named X_train, X_test and Series named y_train, and y_test

assert type(X_train) == pd.DataFrame
assert type(X_test) == pd.DataFrame
assert type(y_train) == pd.Series
assert type(y_test) == pd.Series

# These tests confirm that you have split the data evenly between train and test sets

assert X_train.shape[0] == X_test.shape[0]
assert y_train.shape[0] == y_test.shape[0]


### 4.2) Create an untuned decision tree classifier `wine_dt` and fit it using `X_train` and `y_train`, with `random_state=1`.

Use parameter defaults for your classifier. You must use the Scikit-learn DecisionTreeClassifier (docs [here](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html))

**Starter Code**

    wine_dt =

In [ ]:
# YOUR CODE HERE
#raise NotImplementedError()
from sklearn.tree import DecisionTreeClassifier
wine_dt =DecisionTreeClassifier(random_state=1)
wine_dt.fit(X_train,y_train)



DecisionTreeClassifier(random_state=1)

In [ ]:
# This test confirms that you have created a DecisionTreeClassifier named wine_dt

assert type(wine_dt) == DecisionTreeClassifier

# This test confirms that you have set random_state to 1

assert wine_dt.get_params()['random_state'] == 1

# This test confirms that wine_dt has been fit

sklearn.utils.validation.check_is_fitted(wine_dt)


### 4.3) Create an array `y_pred` generated by using `wine_dt` to make predictions for the test data.

**Starter Code**

    y_pred =

In [ ]:
# YOUR CODE HERE
#raise NotImplementedError()
y_pred =wine_dt.predict(X_test)

In [ ]:
# This test confirms that you have created an array-like object named y_pred

assert type(np.asarray(y_pred)) == np.ndarray


### 4.4) Create a numeric variable `wine_dt_acc` containing the accuracy score for your predictions.

Hint: You can use the `sklearn.metrics` module.

**Starter Code**

    wine_dt_acc =

In [ ]:
# YOUR CODE HERE
#raise NotImplementedError()
wine_dt_acc = accuracy_score(y_test,y_pred)
wine_dt_acc


0.8764044943820225

In [ ]:
# This test confirms that you have created a numeric variable named wine_dt_acc

assert isinstance(wine_dt_acc, Number)


### 4.5) Short Answer: Based on the accuracy score, does the model seem to be performing well or to have substantial performance issues? Explain your answer.

The model is performing well with an accuracy score of 87.6%